typing/typecore.ml

```diff
@@ -169,6 +174,34 @@ type error =
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
+(** Extract a printable name for an expression, if possible. Used in error
+    messages. *)
+let rec texp_has_printable_name =
+  let lid_is_printable lid =
+    try Some (String.concat "." (Longident.flatten lid))
+    with Misc.Fatal_error -> None
+  in
+  let parent expr sep rhs =
+    match texp_has_printable_name expr.exp_desc with
+    | Some expr' -> Some (expr' ^ sep ^ rhs)
+    | None -> None
+  in
+  function
+  | Texp_ident (_, lid, _) -> lid_is_printable lid.txt
+  | Texp_field (expr, lid, _) -> (
+      match lid_is_printable lid.txt with
+      | Some rhs -> parent expr "." rhs
+      | None -> None)
+  | Texp_send (expr, meth) ->
+      let meth' =
+        match meth with
+        | Tmeth_name name -> name
+        | Tmeth_val ident | Tmeth_ancestor (ident, _) -> Ident.name ident
+      in
+      parent expr "#" meth'
+  | Texp_instvar (_, _, name) -> Some name.txt
+  | _ -> None
```

Comment:
I think you can do this with much less code. The problem you have decided to solve is to be able to show the function to the user in the error message, but only if "it can be named", which basically amounts to "it is a short/atomic construct". Is the expression "short" enough that I want to include it in the error message? This is the question.

But you are mixing this the actual printing logic, which is needlessly duplicated. It may also very well be wrong / imperfect; for example, do you handle the case of infix operators properly?

My suggestion: turn this into a `is_short` function that returns a boolean, move it to `typedtree.ml`, and call it from the error message printer (is `funct` short?). For the printing you could call `Untypeast.untype_expression` and then `Pprintast.expression`. (Or maybe you have access to the parsetree directly at error-raising point and don't need `Untypeast`.)

Reply:
I wasn't proud of this code! I've implemented it like you suggested and it's much better. I've included more cases and written down why I choose them.

Reply:
(I don't currently see the new version in the PR.)

Reply:
Sorry, I just pushed it now.